internal/test/scorecard/common_utils.go

```diff
@@ -0,0 +1,387 @@
+							},
+						},
+					},
+				},
+			},
+			GrafanaConfig: &operatorv1beta1.NetworkConfiguration{
+				Annotations: map[string]string{
+					"nginx.ingress.kubernetes.io/backend-protocol": "HTTPS",
+				},
+				IngressSpec: &netv1.IngressSpec{
+					TLS: []netv1.IngressTLS{{}},
+					Rules: []netv1.IngressRule{
+						{
+							Host: "testing.cryostat-grafana",
+							IngressRuleValue: netv1.IngressRuleValue{
+								HTTP: &netv1.HTTPIngressRuleValue{
+									Paths: []netv1.HTTPIngressPath{
+										{
+											Path:     "/",
+											PathType: &pathType,
+											Backend: netv1.IngressBackend{
+												Service: &netv1.IngressServiceBackend{
+													Name: fmt.Sprintf("%s-grafana", name),
+													Port: netv1.ServiceBackendPort{
+														Number: 3000,
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+	}
+	return cr
+}
+
+func createAndWaitTillCryostatAvailable(cr *operatorv1beta1.Cryostat, resources *TestResources) (*operatorv1beta1.Cryostat, error) {
+	client := resources.Client
+	r := resources.TestResult
+
+	cr, err := client.OperatorCRDs().Cryostats(cr.Namespace).Create(context.Background(), cr)
+	if err != nil {
+		logError(r, fmt.Sprintf("failed to create Cryostat CR: %s", err.Error()))
+		return nil, err
+	}
+
+	// Poll the deployment until it becomes available or we timeout
+	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+	defer cancel()
+	err = waitForDeploymentAvailability(ctx, client, cr.Namespace, cr.Name, r)
+	if err != nil {
+		logError(r, fmt.Sprintf("Cryostat main deployment did not become available: %s", err.Error()))
+		return nil, err
+	}
+
+	err = wait.PollImmediateUntilWithContext(ctx, time.Second, func(ctx context.Context) (done bool, err error) {
+		cr, err = client.OperatorCRDs().Cryostats(cr.Namespace).Get(ctx, cr.Name)
+		if err != nil {
+			return false, fmt.Errorf("failed to get Cryostat CR: %s", err.Error())
+		}
+		if len(cr.Status.ApplicationURL) > 0 {
+			return true, nil
+		}
+		r.Log += "application URL is not yet available\n"
+		return false, nil
+	})
+	if err != nil {
+		logError(r, fmt.Sprintf("application URL not found in CR: %s", err.Error()))
+		return nil, err
+	}
+	r.Log += fmt.Sprintf("application is ready at %s\n", cr.Status.ApplicationURL)
+
+	return cr, nil
+}
+
+func waitTillCryostatReachable(base *url.URL, resources *TestResources) error {
+	client := NewHttpClient()
+	r := resources.TestResult
+
+	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+	defer cancel()
+
+	err := wait.PollImmediateUntilWithContext(ctx, time.Second, func(ctx context.Context) (done bool, err error) {
+		url := base.JoinPath("/health")
+		req, err := NewHttpRequest(ctx, http.MethodGet, url.String(), nil)
+		if err != nil {
+			return false, fmt.Errorf("failed to create a Cryostat REST request: %s", err.Error())
+		}
+		req.Header.Add("Accept", "*/*")
```

Comment:
I've noticed a few times where the test appears to proceed before the web server is up and running. The result is a 503 from OpenShift. I would have thought that the deployment availability check and querying the health check like this would prevent that problem.

One option could be to query the `/api/health` Cryostat endpoint instead and parse the results to ensure that the datasource and Grafana are ready as well. What do you think?

Here's the log:
```
--------------------------------------------------------------------------------
Image:      quay.io/ebaron/cryostat-operator-scorecard:recording-01
Entrypoint: [cryostat-scorecard-tests cryostat-recording]
Labels:
    "test":"cryostat-recording"
    "suite":"cryostat"
Results:
    Name: cryostat-recording
    State: fail

    Errors:
        failed to list discovered targets: API request failed with status code 503: <html>
  <head>
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <style type="text/css">
      body {
        font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
        line-height: 1.66666667;
        font-size: 16px;
        color: #333;
        background-color: #fff;
        margin: 2em 1em;
      }
      h1 {
        font-size: 28px;
        font-weight: 400;
      }
      p {
        margin: 0 0 10px;
      }
      .alert.alert-info {
        background-color: #F0F0F0;
        margin-top: 30px;
        padding: 30px;
      }
      .alert p {
        padding-left: 35px;
      }
      ul {
        padding-left: 51px;
        position: relative;
      }
      li {
        font-size: 14px;
        margin-bottom: 1em;
      }
      p.info {
        position: relative;
        font-size: 20px;
      }
      p.info:before, p.info:after {
        content: "";
        left: 0;
        position: absolute;
        top: 0;
      }
      p.info:before {
        background: #0066CC;
        border-radius: 16px;
        color: #fff;
        content: "i";
        font: bold 16px/24px serif;
        height: 24px;
        left: 0px;
        text-align: center;
        top: 4px;
        width: 24px;
      }

      @media (min-width: 768px) {
        body {
          margin: 6em;
        }
      }
    </style>
  </head>
  <body>
    <div>
      <h1>Application is not available</h1>
      <p>The application is currently not serving requests at this endpoint. It may not have been started or is still starting.</p>

      <div class="alert alert-info">
        <p class="info">
          Possible reasons you are seeing this page:
        </p>
        <ul>
          <li>
            <strong>The host doesn't exist.</strong>
            Make sure the hostname was typed correctly and that a route matching this hostname exists.
          </li>
          <li>
            <strong>The host exists, but doesn't have a matching path.</strong>
            Check if the URL path was typed correctly and that the route was created using the desired path.
          </li>
          <li>
            <strong>Route and path matches, but all pods are down.</strong>
            Make sure that the resources exposed by this route (pods, services, deployment configs, etc) have at least one pod running.
          </li>
        </ul>
      </div>
    </div>
  </body>
</html>

        failed to get a target for test: API request failed with status code 503: <html>
  <head>
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <style type="text/css">
      body {
        font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
        line-height: 1.66666667;
        font-size: 16px;
        color: #333;
        background-color: #fff;
        margin: 2em 1em;
      }
      h1 {
        font-size: 28px;
        font-weight: 400;
      }
      p {
        margin: 0 0 10px;
      }
      .alert.alert-info {
        background-color: #F0F0F0;
        margin-top: 30px;
        padding: 30px;
      }
      .alert p {
        padding-left: 35px;
      }
      ul {
        padding-left: 51px;
        position: relative;
      }
      li {
        font-size: 14px;
        margin-bottom: 1em;
      }
      p.info {
        position: relative;
        font-size: 20px;
      }
      p.info:before, p.info:after {
        content: "";
        left: 0;
        position: absolute;
        top: 0;
      }
      p.info:before {
        background: #0066CC;
        border-radius: 16px;
        color: #fff;
        content: "i";
        font: bold 16px/24px serif;
        height: 24px;
        left: 0px;
        text-align: center;
        top: 4px;
        width: 24px;
      }

      @media (min-width: 768px) {
        body {
          margin: 6em;
        }
      }
    </style>
  </head>
  <body>
    <div>
      <h1>Application is not available</h1>
      <p>The application is currently not serving requests at this endpoint. It may not have been started or is still starting.</p>

      <div class="alert alert-info">
        <p class="info">
          Possible reasons you are seeing this page:
        </p>
        <ul>
          <li>
            <strong>The host doesn't exist.</strong>
            Make sure the hostname was typed correctly and that a route matching this hostname exists.
          </li>
          <li>
            <strong>The host exists, but doesn't have a matching path.</strong>
            Check if the URL path was typed correctly and that the route was created using the desired path.
          </li>
          <li>
            <strong>Route and path matches, but all pods are down.</strong>
            Make sure that the resources exposed by this route (pods, services, deployment configs, etc) have at least one pod running.
          </li>
        </ul>
      </div>
    </div>
  </body>
</html>

    Log:
        deployment cryostat-recording is not yet found
        deployment cryostat-recording is not yet found
        deployment cryostat-recording is not yet found
        deployment cryostat-recording is not yet found
        deployment cryostat-recording is not yet found
        deployment cryostat-recording is not yet found
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is not yet available
        deployment cryostat-recording is available
        application is ready at https://cryostat-recording-cryostat-operator-scorecard.apps.example.com/
```

Reply:
ohh I am quite surprised that the issue persists too. My knowledge might be outdated but I can't seem to find the `/api/health` cryostat endpoint. Is it that `/health` endpoint but we add additional check for the response body?

Reply:
```

$ https :8181/api/health --verify=no
HTTP/1.1 404 Not Found
content-encoding: gzip
content-length: 71
content-type: text/html; charset=utf-8

<html><body><h1>Resource not found</h1></body></html>

$ https :8181/health --verify=no
HTTP/1.1 200 OK
content-encoding: gzip
content-length: 138
content-type: application/json

{
    "cryostatVersion": "v2.5.0-dev-95-g64efaac1",
    "dashboardAvailable": true,
    "dashboardConfigured": true,
    "datasourceAvailable": true,
    "datasourceConfigured": true,
    "reportsAvailable": true,
    "reportsConfigured": true
}
```

Reply:
Ah sorry, I'm getting my endpoints mixed up. `/health` is indeed the one we want, `/health/liveness` is the one with fewer checks. We could check that `datasourceAvailable` and `dashboardAvailable` are true. If we do that and we still run into readiness issues, then something weird must be going on.

Reply:
Ahh okayy! Took me a while for crc to launch but I added additional checks by parsing health response and added some missing logs :))